src/schedule/zephyr_dma_domain.c

```diff
@@ -130,57 +141,92 @@ static void zephyr_dma_domain_thread_fn(void *p1, void *p2, void *p3)
 
 static void dma_irq_handler(void *data)
 {
-	struct zephyr_dma_domain_data *zephyr_dma_domain_data;
+	struct zephyr_dma_domain_irq *irq_data;
 	struct zephyr_dma_domain_thread *dt;
-	struct dma_chan_data *channel;
 	struct k_sem *sem;
-	int channel_index, irq;
+	struct list_item *i;
+	struct zephyr_dma_domain_channel *chan_data;
 
-	zephyr_dma_domain_data = data;
-	channel = zephyr_dma_domain_data->channel;
-	channel_index = channel->index;
-	sem = &zephyr_dma_domain_data->dt->sem;
-	irq = zephyr_dma_domain_data->irq;
-	dt = zephyr_dma_domain_data->dt;
+	irq_data = data;
+	sem = &irq_data->dt->sem;
+	dt = irq_data->dt;
+
+	list_for_item(i, &irq_data->channels) {
+		chan_data = container_of(i, struct zephyr_dma_domain_channel, list);
+
+		if (dma_interrupt_legacy(chan_data->channel, DMA_IRQ_STATUS_GET))
+			dma_interrupt_legacy(chan_data->channel, DMA_IRQ_CLEAR);
+	}
 
 	/* clear IRQ */
-	dma_interrupt_legacy(channel, DMA_IRQ_CLEAR);
-	interrupt_clear_mask(irq, BIT(channel_index));
+	interrupt_clear_mask(irq_data->intid, BIT(chan_data->channel->index));
 
 	/* give resources to thread semaphore */
 	if (dt->handler)
 		k_sem_give(sem);
 }
 
-static int enable_dma_irq(struct zephyr_dma_domain_data *data)
+static int enable_dma_irq(struct zephyr_dma_domain_irq *irq_data)
 {
+	struct zephyr_dma_domain_channel *chan_data;
 	int ret;
 
-	dma_interrupt_legacy(data->channel, DMA_IRQ_CLEAR);
+	/* last channel added to irq_data's channel list is the one for
```

Comment:
Do you mean the IRQ enablement step? `enable_dma_irq` is always called during `domain_register`. Since `domain_register` can't be called while another pipeline task is being registered the last channel added to the `channels` list of some IRQ will always be the one for which we need to CLEAR/UNMASK the IRQ. Therefore, the assumption made in the above highlighted code will hold.

Do you have a particular scenario for mixer or aplay || arecord for which you think this may break? It's true I haven't yet  gotten to testing these changes on the xtensa-based platforms.

Reply:
Yes, I was concern of unmasking a "wrong" IRQ.
From your explanations seem it cannot, but, I agree we should test multiple playback/capture in parallel. To be sure we're not missing a (corner) case.